fdbserver/tester.actor.cpp

```diff
@@ -271,104 +272,189 @@ Standalone<VectorRef<KeyValueRef>> checkAllOptionsConsumed(VectorRef<KeyValueRef
-			    .detail("Name", workloadName)
-			    .detail("Count", *wCount)
-			    .detail("Phase", "Start");
-			all.push_back(fmap(
-			    [workloadName, wCount](bool ret) {
-				    --(*wCount);
-				    TraceEvent("WorkloadCheckStatus")
-				        .detail("Name", workloadName)
-				        .detail("Remaining", *wCount)
-				        .detail("Phase", "End");
-				    return true;
-			    },
-			    workloads[i]->check(cx)));
-		}
-		return allTrue(all);
-	}
-
-	ACTOR static Future<std::vector<PerfMetric>> getMetrics(CompoundWorkload* self) {
-		state std::vector<Future<std::vector<PerfMetric>>> results;
-		for (int w = 0; w < self->workloads.size(); w++) {
-			std::vector<PerfMetric> p;
-			results.push_back(self->workloads[w]->getMetrics());
-		}
-		wait(waitForAll(results));
-		std::vector<PerfMetric> res;
-		for (int i = 0; i < results.size(); ++i) {
-			auto const& p = results[i].get();
-			for (auto const& m : p) {
-				res.push_back(m.withPrefix(self->workloads[i]->description() + "."));
-			}
+CompoundWorkload::CompoundWorkload(WorkloadContext& wcx) : TestWorkload(wcx) {}
+
+CompoundWorkload* CompoundWorkload::add(Reference<TestWorkload>&& w) {
+	workloads.push_back(std::move(w));
+	return this;
+}
+
+std::string CompoundWorkload::description() const {
+	std::vector<std::string> names;
+	names.reserve(workloads.size());
+	for (auto const& w : workloads) {
+		names.push_back(w->description());
+	}
+	return fmt::format("{}", fmt::join(std::move(names), ";"));
+}
+Future<Void> CompoundWorkload::setup(Database const& cx) {
+	std::vector<Future<Void>> all;
+	all.reserve(workloads.size());
+	for (int w = 0; w < workloads.size(); w++)
+		all.push_back(workloads[w]->setup(cx));
+	auto done = waitForAll(all);
+	if (failureInjection.empty()) {
+		return done;
+	}
+	std::vector<Future<Void>> res;
+	res.reserve(failureInjection.size());
+	for (auto& f : failureInjection) {
+		res.push_back(f->setupInjectionWorkload(cx, done));
+	}
+	return waitForAll(res);
+}
+Future<Void> CompoundWorkload::start(Database const& cx) {
+	std::vector<Future<Void>> all;
+	all.reserve(workloads.size());
+	auto wCount = std::make_shared<unsigned>(0);
+	for (int i = 0; i < workloads.size(); i++) {
+		std::string workloadName = workloads[i]->description();
+		++(*wCount);
+		TraceEvent("WorkloadRunStatus").detail("Name", workloadName).detail("Count", *wCount).detail("Phase", "Start");
+		all.push_back(fmap(
+		    [workloadName, wCount](Void value) {
+			    --(*wCount);
+			    TraceEvent("WorkloadRunStatus")
+			        .detail("Name", workloadName)
+			        .detail("Remaining", *wCount)
+			        .detail("Phase", "End");
+			    return Void();
+		    },
+		    workloads[i]->start(cx)));
+	}
+	auto done = waitForAll(all);
+	if (failureInjection.empty()) {
+		return done;
+	}
+	std::vector<Future<Void>> res;
+	res.reserve(failureInjection.size());
+	for (auto& f : failureInjection) {
+		res.push_back(f->startInjectionWorkload(cx, done));
+	}
+	return waitForAll(res);
```

Comment:
If `failureInjection` is non-empty, but `waitForAll(res)` is ready before `done` is ready, won't some `TestWorkload::start` actors get cancelled?

Reply:
Good catch, this logic is definitely broken (I actually think it is worse than what you wrote -- it might cancel the workloads immediately). Will post a fix